### PR TITLE
Spherical coordinate support for e57 reader

### PR DESCRIPTION
**Pull request by AkashPushkar**
New feature
- Added support for e57 files in spherical coordinate system
- conversion of spherical coordinates to xyz in read_scan

Minor bug fixes:
In read_scan, invalidCartesianState is used to remove points, only if field is present (https://github.com/davidcaron/pye57/issues/6)

### CODE REVIEW

File: pye57/e57.py [hunk 1]
```diff
@@ -170,20 +194,29 @@ def read_scan(self,
...
         if transform:
-            xyz = np.array([data["cartesianX"], data["cartesianY"], data["cartesianZ"]]).T
-            xyz = self.to_global(xyz, header.rotation, header.translation)
+            if coordinate_system is COORDINATE_SYSTEMS.CARTESIAN:
+                xyz = np.array([data["cartesianX"], data["cartesianY"], data["cartesianZ"]]).T
+                if header.has_pose():
+                    xyz = self.to_global(xyz, header.rotation, header.translation)
+            elif coordinate_system is COORDINATE_SYSTEMS.SPHERICAL:
+                rae = np.array([data["sphericalRange"], data["sphericalAzimuth"], data["sphericalElevation"]]).T
+                # TODO: global tranform for spherical cooridinates
```
Comment (davidcaron) [inline pye57/e57.py:212]: Does this comment refer to the case when a scan has a pose, and is in spherical coordinates? From what I undestand, I think this case is not handled here...

**Review by davidcaron — CHANGES_REQUESTED** [review]
Thank you for this PR.

I think we would need to add a test for this one... maybe convert the `test.e57` file to spherical, and try to read it back.

You can take a look at my other comments too.

File: pye57/scan_header.py [hunk 2]
```diff
@@ -142,6 +146,34 @@ def zMinimum(self):
...
+    def rangeMinimum(self):
+        return self.sphericalBounds["rangeMinimum"].value()
+
+    @property
+    def rangeMaximum(self):
+        return self.sphericalBounds["rangeMaximum"].value()
+
+    @property
+    def elevationMinimum(self):
+        return self.sphericalBounds["rangeMinimum"].value()
```
[the remark this answers is not in the document]
Comment (AkashPushkar) [reply pye57/scan_header.py:163]: nice catch! updated.

File: pye57/e57.py [hunk 3]
```diff
@@ -1,32 +1,50 @@
 import uuid
 import os
-
 from typing import Dict
+from enum import Enum, auto
```
[the remark this answers is not in the document]
Comment (AkashPushkar) [reply pye57/e57.py:4]: done

File: pye57/e57.py (hunk 1 above)
Comment (AkashPushkar) [reply pye57/e57.py:212]: yes, that was the case. But updated it now.

1. `xyz` in calculated in both coordinate system
2. Transformed to global if pose is available

**Comment by AkashPushkar** [comment]
> Thank you for this PR.
> 
> I think we would need to add a test for this one... maybe convert the `test.e57` file to spherical, and try to read it back.
> 
> You can take a look at my other comments too.

Hi David, 

I have made the suggested changes and have added unit test for `read_scan` and `read_header` for spherical case since that is what is changed in the PR.

Instead of using the same test file, I have used test data example (for spherical coordinates) from official site of libe57  (http://www.libe57.org/data.html)

**Comment by chataign** [comment]
@davidcaron Hi David, Akash made the changes your requested, do you need anything else from us to merge this PR? Thanks!

File: src/pye57/utils.py [hunk 4]
```diff
@@ -19,3 +22,20 @@ def get_node(node, name):
...
+    formula is different from the one online, so please use formula at the above reference)
+    """
+    range_ = rae[:, :1]
+    theta = rae[:, 1:2]
+    phi =  rae[:, 2:3]
+    return np.concatenate((
+            range_ * np.cos(phi) * np.cos(theta),
+            range_ * np.cos(phi) * np.sin(theta),
+            range_ * np.sin(phi)
+        ), axis=1)
```
Comment (davidcaron) [inline src/pye57/utils.py:41]:
```suggestion
    range_cos_phi = range_ * np.cos(phi)
    return np.concatenate((
        range_cos_phi * np.cos(theta),
        range_cos_phi * np.sin(theta),
        range_ * np.sin(phi)
    ), axis=1)
```

**Review by davidcaron — APPROVED** [review]
It looks great!

It might be nitpicking, but It seems to be around 15% faster to avoid re-computing range_ * np.cos(phi).

I can make a new release once this is merged.

Thanks

**Comment by chataign** [comment]
thank you !

**Pull request merged**